src/Orleans.Core/Messaging/MessageSerializer.cs

```diff
@@ -149,7 +150,7 @@ private ResponseCodec GetRawCodec(Type fieldType)
             return rawCodec;
         }
 
-        public (int HeaderLength, int BodyLength) Write<TBufferWriter>(ref TBufferWriter writer, Message message) where TBufferWriter : IBufferWriter<byte>
+        public (int HeaderLength, int BodyLength) Write(PipeWriter writer, Message message)
```

Comment:
Could this have been done with a specialized `PipeBufferWriter : IBufferWriter<byte>` struct so that this could be kept generic? I realize that would be extra work for the JIT. I have been considering replacing the networking layer with a new, message-oriented one instead of the current stream-oriented one, which will now require us to change this. I'm ok with that, but just FYI

Reply:
Yes, a specialized struct wrapping `PipeWriter` could work (would get inlined by JIT and not cause generic code to be generated).